File: tests/js_bbt_smoke.mjs
// CPU smoke: load browser IIFEs in Node and compare against a Python fixture.
import fs from "node:fs";
import vm from "node:vm";
import path from "node:path";
import { fileURLToPath } from "node:url";

const here = path.dirname(fileURLToPath(import.meta.url));
const root = path.resolve(here, "..");
const fixturePath = process.argv[2] || path.join(here, "_js_bbt_fixture.json");

const fixture = JSON.parse(fs.readFileSync(fixturePath, "utf8"));
const sandbox = {
  window: {},
  WebGPUBackend: undefined,
  console,
  Float32Array,
  Int32Array,
  Uint32Array,
  Math,
  Error,
  btoa: (s) => Buffer.from(s, "binary").toString("base64"),
  atob: (s) => Buffer.from(s, "base64").toString("binary"),
};
sandbox.window = sandbox;
sandbox.globalThis = sandbox;
vm.createContext(sandbox);
for (const rel of ["gabion/web/tinygrad_v0.js", "gabion/web/bbt_forward.js"]) {
  const src = fs.readFileSync(path.join(root, rel), "utf8");
  vm.runInContext(src, sandbox, { filename: rel });
}

const tg = sandbox.window.tinygradV0;
if (!tg || !tg.BBTTransformer) {
  throw new Error("tinygradV0.BBTTransformer not installed");
}

const cfg = fixture.config;
const model = new tg.BBTTransformer({
  vocabSize: cfg.vocab_size,
  dModel: cfg.d_model,
  nHeads: cfg.n_heads,
  nLayers: cfg.n_layers,
  seqLen: cfg.seq_len,
  dFF: cfg.d_ff,
  tieWeights: cfg.tie_weights,
  actQuant: cfg.act_quant,
});
const weights = Float32Array.from(fixture.weights);
const consumed = model.loadFlatWeights(weights, false);
if (consumed !== weights.length) {
  throw new Error(`weight cursor ${consumed} != ${weights.length} (paramCount=${model.paramCount()})`);
}

const B = fixture.batch_size;
const T = cfg.seq_len;
const xFlat = Int32Array.from(fixture.x_flat);
const yFlat = Int32Array.from(fixture.y_flat);
const logits = await model.forward(xFlat, B, T, !!fixture.ternarize);
const loss = await tg.Tensor.crossEntropy(logits, yFlat);

const jsLogits = Array.from(logits.data);
const jsLoss = Number(loss.data[0]);
const pyLogits = fixture.logits_flat;
if (jsLogits.length !== pyLogits.length) {
  throw new Error(`logit length js=${jsLogits.length} py=${pyLogits.length}`);
}
let maxAbs = 0;
let sumAbs = 0;
for (let i = 0; i < jsLogits.length; i++) {
  const d = Math.abs(jsLogits[i] - pyLogits[i]);
  if (d > maxAbs) maxAbs = d;
  sumAbs += d;
}
const meanAbs = sumAbs / jsLogits.length;
const lossAbs = Math.abs(jsLoss - fixture.loss);

// One CPU train step must finish and keep the vector length.
const trained = await tg.trainLocalV1(weights, {
  lr: 1e-3,
  epochs: 1,
  batchSize: B,
  seed: 1,
  vocabSize: cfg.vocab_size,
  dModel: cfg.d_model,
  nHeads: cfg.n_heads,
  nLayers: cfg.n_layers,
  seqLen: cfg.seq_len,
  dFF: cfg.d_ff,
  tieWeights: cfg.tie_weights,
  ternarize: !!fixture.ternarize,
  sequences: fixture.sequences,
  optimizer: "sgd",
  gradClipNorm: 0,
  warmupSteps: 1,
});
const updated = trained.updated || trained;
const updatedLen = updated.length || trained.weights_count || 0;

const report = {
  ok_load: consumed === weights.length,
  param_count: model.paramCount(),
  js_loss: jsLoss,
  py_loss: fixture.loss,
  loss_abs: lossAbs,
  logits_max_abs: maxAbs,
  logits_mean_abs: meanAbs,
  train_mode: trained.mode || null,
  train_loss: Number(trained.loss),
  train_updated_len: updatedLen,
  finite: Number.isFinite(jsLoss) && Number.isFinite(Number(trained.loss)),
};
const outPath = fixturePath.replace(/\.json$/, "_js.json");
fs.writeFileSync(outPath, JSON.stringify(report, null, 2));
console.log(JSON.stringify(report, null, 2));
console.log(`wrote ${outPath}`);

File: tests/js_bbt_smoke.py
"""CPU numeric smoke: Python BBT vs browser JS port (Node + IIFE stub)."""
from __future__ import annotations

import json
import math
import subprocess
import sys
from pathlib import Path

import numpy as np

ROOT = Path(__file__).resolve().parents[1]
FIXTURE = Path(__file__).resolve().parent / "_js_bbt_fixture.json"


def _flatten(params) -> list[float]:
    from gabion.pebble.adapters import flatten_tensors

    return flatten_tensors(params)


def build_fixture() -> dict:
    from gabion.user_models.bbt_transformer import BBTTransformerAdapter

    cfg = {
        "vocab_size": 32,
        "d_model": 16,
        "n_heads": 2,
        "n_layers": 1,
        "seq_len": 8,
        "d_ff": 32,
        "tie_weights": True,
        "act_quant": True,
    }
    adapter = BBTTransformerAdapter(
        input_dim=cfg["vocab_size"],
        d_model=cfg["d_model"],
        n_heads=cfg["n_heads"],
        n_layers=cfg["n_layers"],
        seq_len=cfg["seq_len"],
        d_ff=cfg["d_ff"],
        act_quant=cfg["act_quant"],
        tie_weights=cfg["tie_weights"],
        use_wikitext=False,
    )
    params = adapter.init_params(seed=7)
    x, y = adapter.sample_batch(batch_size=2, seed=11)
    # Closest match: JS ternarize=true applies STE bitlinear; Python does that
    # on the default (ternarize=False) path.
    logits = adapter.forward(params, x, ternarize=False)
    loss = adapter.loss(logits, y)
    weights = _flatten(params)
    x_np = np.array(x.numpy(), dtype=np.int32)
    y_np = np.array(y.numpy(), dtype=np.int32)
    logits_np = np.array(logits.numpy(), dtype=np.float64)
    sequences = x_np.tolist()
    return {
        "config": cfg,
        "batch_size": int(x_np.shape[0]),
        "ternarize": True,
        "weights": [float(v) for v in weights],
        "x_flat": x_np.reshape(-1).astype(int).tolist(),
        "y_flat": y_np.reshape(-1).astype(int).tolist(),
        "sequences": sequences,
        "logits_flat": logits_np.reshape(-1).astype(float).tolist(),
        "loss": float(loss.item()),
    }


def main() -> int:
    fixture = build_fixture()
    FIXTURE.write_text(json.dumps(fixture), encoding="utf-8")
    print(f"fixture weights={len(fixture['weights'])} logits={len(fixture['logits_flat'])} py_loss={fixture['loss']:.6f}")
    js = Path(__file__).resolve().parent / "js_bbt_smoke.mjs"
    proc = subprocess.run(
        ["node", str(js), str(FIXTURE)],
        cwd=str(ROOT),
        capture_output=True,
        text=True,
    )
    sys.stdout.write(proc.stdout)
    if proc.returncode != 0:
        sys.stderr.write(proc.stderr)
        print(f"node failed exit={proc.returncode}")
        return proc.returncode

    report = json.loads((FIXTURE.with_name("_js_bbt_fixture_js.json")).read_text(encoding="utf-8"))
    # Forward match is the real check. Train step only needs to finish finite.
    logit_ok = report["logits_max_abs"] < 1e-3
    loss_ok = report["loss_abs"] < 1e-4
    train_ok = report["finite"] and report["train_updated_len"] == len(fixture["weights"])
    print(
        "verdict "
        f"logits={'PASS' if logit_ok else 'FAIL'} "
        f"loss={'PASS' if loss_ok else 'FAIL'} "
        f"train={'PASS' if train_ok else 'FAIL'}"
    )
    if not math.isfinite(report["js_loss"]):
        return 1
    if not train_ok:
        return 1
    # Numeric mismatch is reported but still a useful smoke if JS ran.
    return 0 if (logit_ok and loss_ok and train_ok) else 2


if __name__ == "__main__":
    raise SystemExit(main())
